packages/core/src/tools/tool-compatibility/index.ts

```diff
@@ -0,0 +1,384 @@
+          switch (check.kind) {
+            case 'min':
+              if (check.inclusive) {
+                constraints.gte = check.value;
+              } else {
+                constraints.gt = check.value;
+              }
+              break;
+            case 'max':
+              if (check.inclusive) {
+                constraints.lte = check.value;
+              } else {
+                constraints.lt = check.value;
+              }
+              break;
+            case 'multipleOf': {
+              constraints.multipleOf = check.value;
+              break;
+            }
+          }
+        } else {
+          newChecks.push(check);
+        }
+      }
+    }
+    let result = z.number();
+    for (const check of newChecks) {
+      switch (check.kind) {
+        case 'int':
+          result = result.int();
+          break;
+        case 'finite':
+          result = result.finite();
+          break;
+        default:
+          result = result._addCheck(check);
+      }
+    }
+    const description = this.mergeParameterDescription(value.description, constraints);
+    if (description) {
+      result = result.describe(description);
+    }
+    return result as ShapeValue<T>;
+  }
+
+  public defaultZodDateHandler<T extends z.AnyZodObject>(value: z.ZodTypeAny): ShapeValue<T> {
+    const zodDate = value as z.ZodDate;
+    const constraints: DateConstraints = {};
+    const checks = zodDate._def.checks || [];
+    type ZodDateCheck = (typeof checks)[number];
+    const newChecks: ZodDateCheck[] = [];
+    for (const check of checks) {
+      if ('kind' in check) {
+        switch (check.kind) {
+          case 'min':
+            const minDate = new Date(check.value);
+            if (!isNaN(minDate.getTime())) {
+              constraints.minDate = minDate.toISOString();
+            }
+            break;
+          case 'max':
+            const maxDate = new Date(check.value);
+            if (!isNaN(maxDate.getTime())) {
+              constraints.maxDate = maxDate.toISOString();
+            }
+            break;
+          default:
+            newChecks.push(check);
+        }
+      }
+    }
+    constraints.dateFormat = 'date-time';
+    let result = z.string().describe('date-time');
+    const description = this.mergeParameterDescription(value.description, constraints);
+    if (description) {
+      result = result.describe(description);
+    }
+    return result as ShapeValue<T>;
+  }
+
+  public process(tool: ToolToConvert): {
+    description?: string;
+    parameters: Schema;
+  } {
+    if (isVercelTool(tool)) {
+      return {
+        description: tool.description,
+        // TODO: should we also process vercel tool params?
+        parameters: zodSchema(convertVercelToolParameters(tool)),
+      };
+    }
+
+    // Constraints are now embedded in the Zod schema descriptions, so just use the schema as-is
+    const { schema } = this.zodToAISDKSchema(tool.inputSchema);
```

Comment:
Hey team, just an FYI I had an issue with a tool without inputSchema here. It was passed to the underlying zod parser and throws with `Cannot read properties of undefined (reading 'shape')` in the `defaultZodObjectHandler` cc @DanielSLew

Reply:
Thank you! That was fixed here https://github.com/mastra-ai/mastra/pull/4188